src/Data/Graph/ControlFlowVertex.hs

```diff
@@ -99,63 +100,74 @@ instance ToJSON ControlFlowVertex where
 -- Typeclasses to create 'ControlFlowVertex's from 'Term's. Also extracts
 -- 'Name's for terms with symbolic names like Identifiers and Declarations.
 
-class VertexDeclaration syntax where
-  toVertex :: (Declarations1 syntax, Foldable syntax)
-           => Loc
-           -> ModuleInfo
-           -> syntax (Term syntax Loc)
-           -> Maybe (ControlFlowVertex, Name)
+class VertexDeclaration term where
+  toVertex
+    :: ModuleInfo
+    -> term Loc
+    -> Maybe (ControlFlowVertex, Name)
 
-instance (VertexDeclaration' syntax syntax) => VertexDeclaration syntax where
-  toVertex = toVertex'
+instance (VertexDeclaration1 f, Declarations1 f) => VertexDeclaration (Term f) where
+  toVertex info (Term (In a f)) = liftToVertex toVertex a info f
 
-class VertexDeclaration' whole syntax where
-  toVertex' :: (Declarations1 whole, Foldable whole)
-            => Loc
-            -> ModuleInfo
-            -> syntax (Term whole Loc)
-            -> Maybe (ControlFlowVertex, Name)
```

Comment:
1. Any sort of typeclass you want to apply to a term directly recursively should be defined against the term type, not the syntax type; then if necessary we can define a second one for the syntax types (à la `Declarations`/`Declarations1`).

2. Instead of giving this a parameter for the `whole` syntax type, we abstract it over the contained terms completely.

3. Instead of constraining anything on the recursive availability of `toVertex`, we pass in a higher-order function (à la `declaredName`/`liftDeclaredName`).